File: Python/backprojection_2X.py
import numpy as np
import scipy as sp
import pywt

# DOESNT WORK FOR RECTANGULAR IMAGES!!!!!!
def backprojection_2X(Ir, Orig):
    myfilter = 'db2'#'bior4.4'
    upscale_lvl = 1

    mode = 'smooth'
    wd1 = pywt.wavedec2(Orig,myfilter,level=upscale_lvl, mode=mode)
    # corg = np.append(wd1[0].ravel(),[wd1[1][0].ravel(), wd1[1][1].ravel(), wd1[1][2].ravel()])
    ilowc = wd1[0]/2
    rangeImg = [np.amin(ilowc), np.amax(ilowc)]

    Ir = range0toN(Ir, rangeImg)

    wd2 = pywt.wavedec2(Ir, myfilter, level=upscale_lvl, mode=mode)

    crec = wd2
    crec[0] = 2*ilowc
    if(wd2[0].shape != wd2[1][0].shape):
        lst = list()
        for i in range(0,len(wd2[1])):
            # lst.append(wd2[1][i].T)
            lst.append(np.reshape(wd2[1][i].ravel(), wd2[0].shape))
        crec[1] = lst


    irec = pywt.waverec2(crec, myfilter, mode=mode)
    ibp = range0toN(irec, rangeImg)
    return ibp

def range0toN(A, myrange):
    a = myrange[0]
    b = myrange[1]
    # A[:] = [b if ele > b else a if ele < a else ele for ele in A]
    A[A > b] = b
    A[A < a] = a
    return A

File: UFRESH_2X/backprojection_2X.m
function I_bp=backprojection_2X(SR,HR,filter) 
    % the LL part of the HR image is actually our LR image
    [c_org,l_org] = wavedec2(HR,1,filter);
    Ilowc = appcoef2(c_org,l_org,filter,1)./2;
    
    rangeImg=[min(Ilowc(:)) max(Ilowc(:))];
    SR=range0toN(SR,rangeImg);
    [c1,l1] = wavedec2(SR,1,filter);
    [h1,v1,d1] = detcoef2('all',c1,l1,1);
    
    c_rec=[2*Ilowc(:);  h1(:); v1(:); d1(:);];

    I_rec = waverec2(c_rec,l1,filter);

    I_bp=range0toN(I_rec,rangeImg);
end

